src/Service/BillableUnbilledHoursReportService.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace App\Service;
+
+use App\Model\Reports\BillableUnbilledHoursReportData;
+use App\Repository\WorkerRepository;
+use App\Repository\WorklogRepository;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class BillableUnbilledHoursReportService
+{
+    private const SECONDS_TO_HOURS = 1 / 3600;
+
+    public function __construct(
+        private readonly WorklogRepository $worklogRepository,
+        private readonly DateTimeHelper $dateTimeHelper,
+        private readonly WorkerRepository $workerRepository,
+        private readonly TranslatorInterface $translator,
+    ) {
+    }
+
+    public function getBillableUnbilledHoursReport(
+        int $year,
+    ): BillableUnbilledHoursReportData {
+        $billableUnbilledHoursReportData = new BillableUnbilledHoursReportData();
+        ['dateFrom' => $dateFrom, 'dateTo' => $dateTo] = $this->dateTimeHelper->getFirstAndLastDateOfYear($year);
+
+        $billableWorklogs = $this->worklogRepository->findBillableWorklogsByWorkerAndDateRange($dateFrom, $dateTo);
+
+        $projectData = [];
+        $projectTotals = [];
+        $totalHoursForAllProjects = 0;
+
+        foreach ($billableWorklogs as $billableWorklog) {
+            if (false === $billableWorklog->isBilled()) {
```

Comment:
Instead of this you should add a parameter "$isBilled = null" to findBillableWorklogsByWorkerAndDateRange repository method that only applies if $isBilled is not null. That would limit the number of entities you get from the database and thereby limit the memory use and execution time.

Reply:
That's a great idea. I don't see a situation where it is relevant to get is_billed=null from findBillableWorklogsByWorkerAndDateRange at all, so i added a filter for this in the select.

I still want to be able to get billable worklogs, even after they are billed.